pkg/operator/controller/service-dns/controller.go

```diff
@@ -0,0 +1,256 @@
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if len(service.Spec.Selector[gatewayNameLabelKey]) == 0 {
+		log.Info(`service selector has no "`+gatewayNameLabelKey+`" label; reconciliation will be skipped`, "request", request)
+		return reconcile.Result{}, nil
+	}
+
+	var gateway gatewayapiv1beta1.Gateway
+	gatewayName := types.NamespacedName{
+		Namespace: service.Namespace,
+		Name:      service.Spec.Selector[gatewayNameLabelKey],
+	}
+	if err := r.cache.Get(ctx, gatewayName, &gateway); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("gateway not found; reconciliation will be skipped", "request", request)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	domains := getGatewayHostnames(&gateway)
+	var errs []error
+	errs = append(errs, r.ensureDNSRecordsForGateway(ctx, &gateway, &service, domains.List())...)
+	errs = append(errs, r.deleteStaleDNSRecordsForGateway(ctx, &gateway, &service, domains)...)
+	return reconcile.Result{}, utilerrors.NewAggregate(errs)
+}
+
+// getGatewayHostnames returns a sets.String with the hostnames from the given
+// gateway's listeners.
+func getGatewayHostnames(gateway *gatewayapiv1beta1.Gateway) sets.String {
+	domains := sets.NewString()
+	for _, listener := range gateway.Spec.Listeners {
+		if listener.Hostname == nil || len(*listener.Hostname) == 0 {
+			continue
+		}
+		domains.Insert(string(*listener.Hostname))
+	}
+	return domains
+}
+
+// ensureDNSRecordsForGateway ensures that a DNSRecord CR exists, associated
+// with the given gateway and service, for each of the given domains.  It
+// returns a list of any errors that result from ensuring those DNSRecord CRs.
+func (r *reconciler) ensureDNSRecordsForGateway(ctx context.Context, gateway *gatewayapiv1beta1.Gateway, service *corev1.Service, domains []string) []error {
+	labels := map[string]string{
+		gatewayNameLabelKey: gateway.Name,
+	}
+	for k, v := range service.Labels {
+		labels[k] = v
+	}
+	ownerRef := metav1.OwnerReference{
+		APIVersion: corev1.SchemeGroupVersion.String(),
+		Kind:       "Service",
+		Name:       service.Name,
+		UID:        service.UID,
+	}
+	var errs []error
+	for _, domain := range domains {
+		name := operatorcontroller.GatewayDNSRecordName(gateway, domain)
+		_, _, err := dnsrecord.EnsureDNSRecord(r.client, name, labels, ownerRef, domain, service)
+		errs = append(errs, err)
+	}
+	return errs
+}
+
+// deleteStaleDNSRecordsForGateway deletes any DNSRecord CRs that are associated
+// with the given gateway but specify a DNS name that is not in the given set of
+// domains.  Such DNSRecord CRs may exist if a hostname was modified or deleted
+// on the gateway.  deleteStaleDNSRecordsForGateway returns a list of any errors
+// that result from deleting those DNSRecord CRs.
+func (r *reconciler) deleteStaleDNSRecordsForGateway(ctx context.Context, gateway *gatewayapiv1beta1.Gateway, service *corev1.Service, domains sets.String) []error {
```

Comment:
*nit* `sets.String` is supposedly deprecated. Goland says you should use a generic set instead.

Reply:
Sounds good, we can do it later.